File: src/main/java/no/fintlabs/consumer/config/CaffeineCacheConfig.java
package no.fintlabs.consumer.config;

import com.github.benmanes.caffeine.cache.Cache;
import com.github.benmanes.caffeine.cache.Caffeine;
import no.fintlabs.adapter.models.event.RequestFintEvent;
import no.fintlabs.adapter.models.event.ResponseFintEvent;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

import java.util.concurrent.TimeUnit;

@Configuration
public class CaffeineCacheConfig {

    @Bean
    public Cache<String, String> stringCache() {
        return Caffeine.newBuilder()
                .expireAfterWrite(4, TimeUnit.HOURS)
                .build();
    }

    @Bean
    public Cache<String, ResponseFintEvent> responseFintEvents() {
        return Caffeine.newBuilder()
                .expireAfterWrite(4, TimeUnit.HOURS)
                .build();
    }

}

File: src/main/java/no/fintlabs/consumer/config/CaffeineCacheConfig.kt
package no.fintlabs.consumer.config

import com.github.benmanes.caffeine.cache.Cache
import com.github.benmanes.caffeine.cache.Caffeine
import no.fintlabs.adapter.models.event.ResponseFintEvent
import no.fintlabs.autorelation.buffer.RelationKey
import no.novari.fint.model.resource.Link
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import java.util.concurrent.TimeUnit

@Configuration
open class CaffeineCacheConfig {
    @Bean
    open fun stringCache(): Cache<String, String> =
        Caffeine
            .newBuilder()
            .expireAfterWrite(4, TimeUnit.HOURS)
            .build()

    @Bean
    open fun responseFintEvents(): Cache<String, ResponseFintEvent> =
        Caffeine
            .newBuilder()
            .expireAfterWrite(4, TimeUnit.HOURS)
            .build()

    // TODO: Move directly to UnresolvedRelationCache & make ttl configurable
    @Bean
    open fun relationLinkCache(): Cache<RelationKey, MutableList<Link>> =
        Caffeine
            .newBuilder()
            .expireAfterWrite(7, TimeUnit.DAYS)
            .build()
}
